src/conn/conn_api.c

```diff
@@ -1436,11 +1440,19 @@ __conn_write_config(
 		WT_ERR_NOTFOUND_OK(ret);
 	}
 
-err:	WT_TRET(fclose(fp));
+	/* Flush to disk and close our handle. */
+	WT_ERR(__wt_fp_close(session, &fp));
 
-	/* Don't leave a damaged file in place. */
-	if (ret != 0)
-		(void)__wt_remove(session, filename);
+	/*
+	 * Rename the file into place: that's an atomic operation so we can be
+	 * confident we never see a partial file.
+	 */
+	WT_ERR(__wt_rename(session, WT_BASECONFIG_SET, WT_BASECONFIG));
+
+err:	WT_TRET(__wt_fp_close(session, &fp));
+
+	/* Discard any damaged temporary file, not required but cleaner. */
+	WT_TRET(__wt_remove_if_exists(session, WT_BASECONFIG_SET));
```

Comment:
fclose implies fflush, but fflush does not imply fsync. If this is a file
that is expected to be durable once created (e.g. not to be 0-length after
a crash) then an fsync is needed as well.

On Fri, Mar 20, 2015 at 2:24 PM, Keith Bostic <notifications@github.com>
wrote:

> In src/conn/conn_api.c
> <https://github.com/wiredtiger/wiredtiger/pull/1785#discussion_r26863773>:
>
> >
> > -	/* Don't leave a damaged file in place. */
> > -	if (ret != 0)
> > -		(void)__wt_remove(session, filename);
> > +	/*
> > +	 * Rename the file into place: that's an atomic operation so we can be
> > +	 * confident we never see a partial file.
> > +	 */
> > +	WT_ERR(__wt_rename(session, WT_BASECONFIG_SET, WT_BASECONFIG));
> > +
> > +err:	WT_TRET(__wt_fp_close(session, &fp));
> > +
> > +	/* Discard any damaged temporary file, not required but cleaner. */
> > +	WT_TRET(__wt_remove_if_exists(session, WT_BASECONFIG_SET));
>
> Not sure about the single routine -- it's intriguing, I'll take a run at
> it.
>
> No separate sync is required before the close, fclose implies fflush.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wiredtiger/wiredtiger/pull/1785/files#r26863773>.
>